core/src/main/scala/org/apache/spark/sql/streaming/SnappySinkCallback.scala

```diff
@@ -0,0 +1,184 @@
+import org.apache.spark.util.Utils
+
+/**
+ * Should be implemented by clients who wants to override default behavior provided by
+ * [[DefaultSnappySinkCallback]].
+ * <p>
+ * In order to override the default callback behavior the qualified name of the implementing
+ * class needs to be passed against `sinkCallback` option while defining stream query.
+ *
+ */
+trait SnappySinkCallback {
+
+  /**
+   * This method is called for each streaming batch after checking the possibility of batch
+   * duplication which is indicated by `possibleDuplicate` flag.
+   * <p>
+   * A duplicate batch might be picked up for processing in case of failure. In case of batch
+   * duplication, this method should handle batch in idempotent manner in order to avoid
+   * data inconsistency.
+   */
+  def process(snappySession: SnappySession, sinkProps: Map[String, String],
+              batchId: Long, df: Dataset[Row], possibleDuplicate: Boolean = false): Unit
+}
+
+class SnappyStoreSinkProvider extends StreamSinkProvider with DataSourceRegister {
+
+  @Override
+  def createSink(
+                  sqlContext: SQLContext,
+                  parameters: Map[String, String],
+                  partitionColumns: Seq[String],
+                  outputMode: OutputMode): Sink = {
+    createSinkStateTableIfNotExist(sqlContext)
+    val cc = try {
+      Utils.classForName(parameters(SINK_CALLBACK)).newInstance()
+    } catch {
+      case _: NoSuchElementException => new DefaultSnappySinkCallback()
+    }
+
+    SnappyStoreSink(sqlContext.asInstanceOf[SnappyContext].snappySession, parameters,
+      cc.asInstanceOf[SnappySinkCallback])
+  }
+
+  private def createSinkStateTableIfNotExist(sqlContext: SQLContext) = {
+    sqlContext.asInstanceOf[SnappyContext].snappySession.sql(s"create table if not exists" +
+      s" $SINK_STATE_TABLE (" +
+      " stream_query_id varchar(200)," +
+      " batch_id long, " +
+      " PRIMARY KEY (stream_query_id)) using row options(DISKSTORE 'GFXD-DD-DISKSTORE')")
+  }
+
+  @Override
+  def shortName(): String = SNAPPY_SINK_NAME
+
+}
+
+case class SnappyStoreSink(snappySession: SnappySession,
+    parameters: Map[String, String], sinkCallback: SnappySinkCallback) extends Sink {
+
+  override def addBatch(batchId: Long, data: Dataset[Row]): Unit = {
+    val streamQueryId = parameters(STREAM_QUERY_ID).toUpperCase
+
+    val updated = snappySession.sql(s"update $SINK_STATE_TABLE " +
+      s"set batch_id=$batchId where stream_query_id='$streamQueryId' and batch_id != $batchId")
+      .collect()(0).getAs("count").asInstanceOf[Long]
+
+    // TODO: use JDBC connection here
+    var posDup = false
+
+    if (updated == 0) {
+      try {
+        snappySession.insert(SINK_STATE_TABLE, Row(streamQueryId, batchId))
+        posDup = false
+      }
+      catch {
+        case e: SQLException if e.getSQLState.equals("23505") => posDup = true
+        case exception: Throwable => throw exception
```

Comment:
Don't need to explicitly catch and throw exception. Just remove the line and that will happen either way.